### PR TITLE
Add ARM CPU plugin module

### PR DESCRIPTION
**Pull request by alvoron**
With this PR we would like to introduce OpenVINO plugin which enable to run inference on ARM CPU while using the same OpenVINO API as on Intel platforms. The plugin is built as a wrapper over [ARM Compute Library](https://github.com/ARM-software/ComputeLibrary), which implements low-level NN layers and provides framework to build inference application on ARM system.

### CODE REVIEW

File: modules/arm_plugin/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Copyright (C) 2018 Intel Corporation
```
Comment (ilya-lavrenov) [inline modules/arm_plugin/CMakeLists.txt:1]: Let's update years in all files including source code. I propose `2020-2021`

File: modules/arm_plugin/src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    "${IE_MAIN_ARM_PLUGIN_SOURCE_DIR}/include")
+
+target_link_libraries(${TARGET_NAME} PRIVATE
+    IE::inference_engine
+    arm_compute::arm_compute
+    arm_compute::arm_compute_core
+    arm_compute::half
+    ${NGRAPH_LIBRARIES}
+    IE::inference_engine_transformations
+    ${INTEL_ITT_LIBS}
```
Comment (ilya-lavrenov) [inline modules/arm_plugin/src/CMakeLists.txt:35]: do you use this library directly? I think we don't need it since ITT comes via `IE::inference_engine_plugin_api`

File: modules/arm_plugin/src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# Copyright (C) 2020 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+set(TARGET_NAME "armPlugin")
+
+set(CMAKE_CXX_STANDARD 14)
```
Comment (ilya-lavrenov) [inline modules/arm_plugin/src/CMakeLists.txt:7]: do you use 14 standard features?

File: modules/mo_pytorch/setup.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
```
Comment (l-bat) [inline modules/mo_pytorch/setup.py:1]: Is this file needed in this PR?

Comment (alvoron) [reply modules/mo_pytorch/setup.py:1]: Dmitry already created a separated PR for these mo_pytorch changes: https://github.com/openvinotoolkit/openvino_contrib/pull/18
I applied these changes in advance to pass some security scans. 
@dkurt would you like me to remove your changes from my PR?

**Comment by dkurt** [comment]
Please disable faled test from mo_pytorch module: https://github.com/openvinotoolkit/openvino_contrib/blob/master/modules/mo_pytorch/test/run_tests.py
```
======================================================================
ERROR: test_inception_v3 (__main__.TestModels)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_tests.py", line 93, in test_inception_v3
    self.check_torchvision_model(models.inception_v3, (299, 299), 4e-5)
  File "run_tests.py", line 79, in check_torchvision_model
    mo_pytorch.convert(model, input_shape=inp_size, model_name='model')
  File "/agent/_work/1/openvino_contrib/modules/mo_pytorch/mo_pytorch.py", line 120, in convert
    raise Exception('model conversion failed')
Exception: model conversion failed
```
@likholat, please take a look

File: modules/mo_pytorch/setup.py (hunk 4 above)
Comment (dkurt) [reply modules/mo_pytorch/setup.py:1]: That's for SDL process checks (originally from https://github.com/openvinotoolkit/openvino_contrib/pull/18)

File: modules/arm_plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+mkdir build && cd build
+```
+6. Build OpenVINO™ with ARM plugin:
+```
+ cmake -DCMAKE_BUILD_TYPE=Release \
+       -DCMAKE_TOOLCHAIN_FILE="../cmake/arm.toolchain.cmake" \
+       -DTHREADS_PTHREAD_ARG="-pthread" \
+       -DIE_EXTRA_MODULES=/openvino_contrib/modules \
+       -DBUILD_java_api=OFF \
+       -DBUILD_mo_pytorch=OFF .. && make
```
Comment (dkurt) [inline modules/arm_plugin/README.md:72]: mo_pytorch is not CMake based module

File: modules/mo_pytorch/setup.py (hunk 4 above)
Comment (dkurt) [reply modules/mo_pytorch/setup.py:1]: I never mind to keep it in this PR. Changes are trivial

File: modules/arm_plugin/README.md (hunk 5 above)
Comment (alvoron) [reply modules/arm_plugin/README.md]: removed `-DBUILD_mo_pytorch=OFF` from the command

File: modules/arm_plugin/CMakeLists.txt (hunk 1 above)
Comment (alvoron) [reply modules/arm_plugin/CMakeLists.txt]: fixed

File: modules/arm_plugin/src/CMakeLists.txt (hunk 3 above)
Comment (apankratovantonp) [reply modules/arm_plugin/src/CMakeLists.txt:7]: Yes

File: modules/arm_plugin/Dockerfile.RPi32 [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ENV SOURCE_UPDATES_OF 20210120_11
+
+#OpenCV
+RUN git clone https://github.com/opencv/opencv.git $OPENCV_HOME && \
+    mkdir $OPENCV_HOME/build && \
+    cd $OPENCV_HOME/build && \
+    cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_LIST=imgcodecs,videoio,highgui \
+             -DCMAKE_TOOLCHAIN_FILE=$OPENCV_HOME/platforms/linux/arm-gnueabi.toolchain.cmake \
+             -DCMAKE_STAGING_PREFIX=$STAGING_DIR && \
+    make -j32 && \
```
Comment (alalek) [inline modules/arm_plugin/Dockerfile.RPi32:106]: > -j32

This is bad practice to oversubscribe resources.
Use `$(nproc)` for that or use Docker's `--build-arg` with some default (e.g. 4).

File: modules/arm_plugin/Dockerfile.RPi32 [hunk 7]
```diff
@@ -0,0 +1,137 @@
+FROM debian:9
+
```
Comment (alalek) [inline modules/arm_plugin/Dockerfile.RPi32:2]: It makes sense to mention that this Dockerfile is for x86_64 arch and should be used for cross-compilation.
(not for ARM platforms)

File: modules/arm_plugin/Dockerfile.RPi32 [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+# build and install git 2.22
+RUN wget https://github.com/git/git/archive/v2.22.2.tar.gz && \
+    tar xf v2.22.2.tar.gz && \
+    cd git-2.22.2 && \
+    make configure && \
+    ./configure --prefix=/usr && \
+    make all && \
+    make install && \
+    rm -rf /git-2.22.2 && \
+    rm /v2.22.2.tar.gz
```
Comment (alalek) [inline modules/arm_plugin/Dockerfile.RPi32:85]: Comments: what? => why?

Please add comment describing why system "git" package version is not enough (which features are required?)

File: modules/arm_plugin/Dockerfile.RPi32 [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+ENV OPENCV_HOME=$DEV_HOME/opencv
+ENV OPENVINO_HOME $DEV_HOME/openvino
+ENV OPENVINO_CONTRIB $DEV_HOME/openvino_contrib
+ENV ARM_PLUGIN_HOME $OPENVINO_CONTRIB/modules/arm_plugin
+ENV STAGING_DIR $DEV_HOME/armcpu_debian32_rpi
+
+ENV SOURCE_UPDATES_OF 20210120_11
+
+#OpenCV
+RUN git clone https://github.com/opencv/opencv.git $OPENCV_HOME && \
```
Comment (alalek) [inline modules/arm_plugin/Dockerfile.RPi32:100]: How is automatic Docker's cache of built images handled here?

File: modules/arm_plugin/Dockerfile.RPi32 [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        libffi-dev \
+        libssl-dev \
+        chrpath \
+        libssl-dev \
+        libprotobuf-dev \
+        libprotoc-dev \
+        protobuf-compiler \
+        # For ARM CPU plugin
+        scons
+
```
Comment (alalek) [inline modules/arm_plugin/Dockerfile.RPi32:62]: Follow BKMs: cleanup apt caches

File: modules/arm_plugin/README.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+ ```
+cd openvino_contrib/modules/arm_plugin
+```
+3. Build the plugin in Docker container:
+```
+docker image build -t arm-plugin -f Dockerfile.RPi32 .
+```
+4. Export the archive with artifacts to the current directory:
+```
+docker run -ti -v $PWD:/remote arm-plugin cp ./OV_ARM_package.tar.gz /remote
```
Comment (alalek) [inline modules/arm_plugin/README.md:34]: - `docker run --rm`
- do you need "root"-owned files? what is about files permissions?

File: modules/arm_plugin/README.md [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+1. Clone `openvino` and `openvino_contrib` repositories:
+```
+git clone --recurse-submodules --single-branch --branch=master https://github.com/openvinotoolkit/openvino.git 
+git clone --recurse-submodules --single-branch --branch=master https://github.com/openvinotoolkit/openvino_contrib.git 
+```
+2. Run Docker container with mounted both `openvino` and `openvino_contrib` repositories if you do cross-compilation. If you do native compilation just skip this step:
+```
+docker run -it -v /absolute/path/to/openvino:/openvino -v /absolute/path/to/openvino_contrib:/openvino_contrib ie_cross_armhf /bin/bash 
+```
+The next commands of the procedure need to be run in `ie_cross_armhf` container.
```
Comment (alalek) [inline modules/arm_plugin/README.md:52]: > ie_cross_armhf

Any reference with details?

File: modules/arm_plugin/README.md [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+```
+3. Download a vehicle image, for instance, [this image]:
+```
+wget https://raw.githubusercontent.com/openvinotoolkit/openvino/master/scripts/demo/car_1.bmp
+```
+4. Copy model [Intermediate Representation] (`vehicle-license-plate-detection-barrier-0123.bin` and `vehicle-license-plate-detection-barrier-0123.xml` files generated by model converter) to ARM platform.
+
+5. Run object detection sample on ARM platform:
+```
+export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:lib
```
Comment (alalek) [inline modules/arm_plugin/README.md:142]: It is not recommended to use relative paths in `LD_LIBRARY_PATH` from security perspective:

```
$(pwd)/lib
```

File: modules/arm_plugin/README.md [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+./object_detection_sample_ssd -m vehicle-license-plate-detection-barrier-0123.xml -i car_1.bmp -d ARM
+```
+
+On the output image you should see 2 cars enclosed in purple rectangles and a front plate enclosed in grey rectangle:
+
+![](https://user-images.githubusercontent.com/1412335/103134082-83458000-46bf-11eb-90b5-ef3b7ccd23ff.jpg)
```
Comment (alalek) [inline modules/arm_plugin/README.md:148]: Why is this image not a part of repository?

File: modules/arm_plugin/thirdparty/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,170 @@
+# Copyright (C) 2020-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if(DEFINED ARM_COMPUTE_INCLUDE_DIR OR DEFINED ARM_COMPUTE_LIB_DIR)
+    if (NOT DEFINED ARM_COMPUTE_INCLUDE_DIR)
+        message(STATUS "Undefined ARM_COMPUTE_INCLUDE_DIR input variable should be set manually")
```
Comment (alalek) [inline modules/arm_plugin/thirdparty/CMakeLists.txt:7]: > STATUS

`FATAL_ERROR`?

File: modules/java_api/build.gradle [hunk 16]
```diff
@@ -0,0 +1,28 @@
+plugins {
```
Comment (alalek) [inline modules/java_api/build.gradle:1]: Is this file related to this PR?

**Review by alalek — COMMENTED** [review]
Well done!

Some feedback notes are below.

File: modules/mo_pytorch/setup.py (hunk 4 above)
Comment (alalek) [reply modules/mo_pytorch/setup.py:1]: Perhaps git rebase should be applied to keep patches and history clean.

File: modules/arm_plugin/README.md (hunk 12 above)
Comment (alvoron) [reply modules/arm_plugin/README.md:52]: @ilya-lavrenov do we have some description of `ie_crocc_armhf` image? I've found some details in the issue discussion https://github.com/openvinotoolkit/openvino/issues/500

File: modules/java_api/build.gradle (hunk 16 above)
Comment (alvoron) [reply modules/java_api/build.gradle:1]: This file is needed to pass security checks required to release the plugin.

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 8 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32:85]: If system git package is used than rate limit exceeded when openvino repo is being cloned:
```
Error downloading object: docs/img/benefit_ncs2.png (ba09c1c): Smudge error: Error downloading docs/img/benefit_ncs2.png (ba09c1c8b748936350df0999b1a926a41ca040f171754bd7884950fe13402e4c): batch response: Rate limit exceeded: https://github.com/openvinotoolkit/openvino.git/info/lfs/objects/batch
```
I don't know what exactly feature\fix of git 2.22.2 solves this issue. It's better to ask @terfendail but he's on sick leave.

File: modules/arm_plugin/README.md (hunk 14 above)
Comment (alvoron) [reply modules/arm_plugin/README.md:148]: This image is output of Object Detection for OpenVINO SSD sample: https://docs.openvinotoolkit.org/latest/openvino_inference_engine_samples_object_detection_sample_ssd_README.html
I'm not sure we need to store output of OpenVINO sample in contrib repository.

File: modules/arm_plugin/src/CMakeLists.txt (hunk 2 above)
Comment (alvoron) [reply modules/arm_plugin/src/CMakeLists.txt]: removed `${INTEL_ITT_LIBS}`

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 6 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32]: set `make -j$(nproc)` in `Dockerfile.RPi32` and `Dockerfile.RPi63`

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 7 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32:5]: added a comment to `Dockerfile.RPi32` and `Dockerfile.RPi64`

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 10 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32]: added apt clean cache command

File: modules/arm_plugin/README.md (hunk 13 above)
Comment (alvoron) [reply modules/arm_plugin/README.md]: fixed the path

File: modules/arm_plugin/thirdparty/CMakeLists.txt (hunk 15 above)
Comment (alvoron) [reply modules/arm_plugin/thirdparty/CMakeLists.txt:7]: @l-bat @apankratovantonp could you please comment this question?

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 9 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32:104]: As I know, there is no way to disable Docker cache for particular commands of Dockerfile. I think we shouldn't add `--no-cache` option to `docker run` command by default, but we need to add a note to README that `--no-cache` option should be used to get the latest changes from opencv\openvino\openvino_contrib repositories.

**Review by vladimir-dudnik — APPROVED** [review]

**Review by l-bat — APPROVED** [review]

File: modules/arm_plugin/README.md (hunk 12 above)
Comment (ilya-lavrenov) [reply modules/arm_plugin/README.md:52]: I don't know about IE docker images..

@dkurt do you know anything about this image?

Comment (alalek) [reply modules/arm_plugin/README.md:52]: > ie_cross_armhf

This was mentioned in `build-instruction.md` file (openvino repository).
Currently this file was gone (from the master branch) or moved somewhere.
2020.3 LTS branch: https://github.com/openvinotoolkit/openvino/blob/releases/2020/3/build-instruction.md

**Review by ilya-lavrenov — APPROVED** [review]

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 9 above)
Comment (terfendail) [reply modules/arm_plugin/Dockerfile.RPi32:104]: As far as I could find out Docker doesn't invalidate caches on repository update. That's why I use `ENV SOURCE_UPDATE_OF` to retrigger evaluation of following steps.

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 8 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32]: It was found that system `git lfs` solves this issue. Removed git 2.22 from Dockerfile and replace `git` with `git lfs` to clone OpenVINO repository.

File: modules/arm_plugin/Dockerfile.RPi32 (hunk 9 above)
Comment (alvoron) [reply modules/arm_plugin/Dockerfile.RPi32:94]: Added the note about `--no-cache` option to `README.md`.

File: modules/arm_plugin/README.md (hunk 11 above)
Comment (alvoron) [reply modules/arm_plugin/README.md]: Added `--rm` option to `docker run` command.
Owner of copied `OV_ARM_package.tar.gz` archive is root but file permissions are 644 which allows to read the archive by non-root accounts.

**Review by terfendail — APPROVED** [review]

**Review by alalek — APPROVED** [review]
Looks good to me :+1:

File: modules/arm_plugin/thirdparty/CMakeLists.txt (hunk 15 above)
Comment (apankratovantonp) [reply modules/arm_plugin/thirdparty/CMakeLists.txt]: Yes FATAL_ERROR should be used

File: modules/arm_plugin/README.md (hunk 12 above)
Comment (dkurt) [reply modules/arm_plugin/README.md:55]: Hi! This is from https://github.com/openvinotoolkit/openvino/blob/releases/2021/1/build-instruction.md#cross-compilation-using-docker
